include/hal/ibex/ibex.h

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright (C) 2020 ETH Zurich and University of Bologna
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __HAL_IBEX_IBEX_H__
+#define __HAL_IBEX_IBEX_H__
+
+#include "archi/pulp.h"
+
+#include "archi/riscv/builtins_v2_emu.h"
+
+// For PULP, ibex added non-standard irqs to allow for 32 fast interrupts. 
+// These use custom CSRs: MIE: 0x7D0, MTVEC: 0x7D1, MIP: 0x7D2
+#define SR_MTVEC  0x7D1
+
+#define hal_spr_read_then_clr(reg,val) \
+  ({ \
+    int state; \
+    asm volatile ("csrrc %0, %1, %2" :  "=r" (state) : "I" (reg), "I" (val) ); \
+    state; \
+  })
```

Comment:
I guess at some point I will make this file the global default.

There is literally zero reason why we have builtins to read csr registers I don't know who thought this is a good idea.

Anyway this is good stuff